File: backend/api/export.py
"""
Export API Endpoints

Handles exporting analysis results to CSV format.
"""

import csv
import io
import logging
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import List, Dict, Any, Optional

logger = logging.getLogger(__name__)

router = APIRouter()


class ExportRequest(BaseModel):
    """Request model for exporting results."""
    data: List[Dict[str, Any]]
    filename: Optional[str] = "export.csv"
    columns: Optional[List[str]] = None


@router.post("/results")
async def export_results(request: ExportRequest):
    """
    Export analysis results as CSV file.

    Args:
        request: ExportRequest with data, filename, and optional column order

    Returns:
        StreamingResponse with CSV file
    """
    try:
        if not request.data:
            raise HTTPException(status_code=400, detail="No data provided for export")

        # Create CSV in memory
        output = io.StringIO()

        # Determine columns from the data
        if request.columns:
            fieldnames = request.columns
        else:
            # Get all unique keys from all rows
            fieldnames = list(dict.fromkeys(
                key for row in request.data for key in row.keys()
            ))

        writer = csv.DictWriter(output, fieldnames=fieldnames, extrasaction='ignore')
        writer.writeheader()

        for row in request.data:
            writer.writerow(row)

        # Reset position to start of StringIO
        output.seek(0)

        # Create response
        filename = request.filename if request.filename.endswith('.csv') else f"{request.filename}.csv"

        logger.info(f"Exporting {len(request.data)} rows to {filename}")

        return StreamingResponse(
            iter([output.getvalue()]),
            media_type="text/csv",
            headers={
                "Content-Disposition": f"attachment; filename={filename}",
                "Cache-Control": "no-cache"
            }
        )

    except Exception as e:
        logger.error(f"Error exporting results: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to export results: {str(e)}")

File: frontend/src/api/export.ts
/**
 * Export API Client
 *
 * Handles exporting analysis results to CSV files
 */

const API_BASE_URL = '/api'

interface ExportRequest {
  data: Array<Record<string, any>>
  filename?: string
  columns?: string[]
}

/**
 * Export data as CSV file
 * @param data Array of data objects to export
 * @param filename Optional filename (default: "export.csv")
 * @param columns Optional column order
 */
export async function exportResultsAsCSV(
  data: Array<Record<string, any>>,
  filename?: string,
  columns?: string[]
): Promise<void> {
  const requestBody: ExportRequest = {
    data,
    filename: filename || 'export.csv',
    columns
  }

  const response = await fetch(`${API_BASE_URL}/export/results`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(requestBody),
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to export data' }))
    throw new Error(error.detail || 'Failed to export data')
  }

  // Get the blob from response
  const blob = await response.blob()

  // Create a download link and trigger download
  const url = window.URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename || 'export.csv'
  document.body.appendChild(link)
  link.click()

  // Cleanup
  document.body.removeChild(link)
  window.URL.revokeObjectURL(url)
}
